nimble/host/src/ble_l2cap_coc.c

```diff
@@ -88,17 +103,81 @@ ble_l2cap_coc_create_server(uint16_t psm, uint16_t mtu,
     return 0;
 }
 
-static uint16_t
-ble_l2cap_coc_get_cid(void)
+static int
+ble_l2cap_coc_alloc_cid(cid_mask_t *cid_chan_mask)
+{
+    int i, j;
+    /* The following equation is used to get the Maximum Value of bitmask
+     * cid_mask = ((1UL << TOTAL_CID_PER_MASK_ELEMENT) - 1)
+     * but using -1 is more generic in terms of cid_mask_t as all octets
+     * must be 0xFF
+     * cid_mask is used to search for the first free slot to use it.
+     * */
+    cid_mask_t cid_mask = -1;
+    uint16_t cid_idx;
+
+    for(i = 0; i < CID_MASK_LEN; i++) {
+        if(cid_chan_mask[i] != cid_mask) {
+             break;
+        }
+    }
+
+    if (i == CID_MASK_LEN) {
+        /* No Free CID */
+        return -1;
+    }
+
+    cid_mask = cid_chan_mask[i];
+
+    for(j = 0; j < TOTAL_CID_PER_MASK_ELEMENT; j++ ) {
+        if (cid_mask & 0x01) {
+            cid_mask >>= 1;
+        }
+        else {
+             break;
+        }
+    }
+    if (j == TOTAL_CID_PER_MASK_ELEMENT) {
+        /* Can't happen */
+        return -1;
+    }
+    cid_idx = (i * TOTAL_CID_PER_MASK_ELEMENT) + j;
+    /* Set CID in Mask */
+    SET_CID_MASK(cid_chan_mask,cid_idx);
+
+    return cid_idx;
+}
+
+static void
+ble_l2cap_coc_free_cid(struct ble_hs_conn *conn, uint16_t cid)
+{
+	BLE_HS_DBG_ASSERT((cid >= BLE_L2CAP_COC_CID_START) &&
+                      (cid <= BLE_L2CAP_COC_CID_END));
+    if(conn) {
+        cid -= BLE_L2CAP_COC_CID_START;
+        RESET_CID_MASK(conn->bhc_cid_mask, cid);
+    }
+}
+
+static int
+ble_l2cap_coc_get_cid(struct ble_hs_conn *conn)
 {
-    static uint16_t next_cid = BLE_L2CAP_COC_CID_START;
+    int next_cid;
 
-    if (next_cid > BLE_L2CAP_COC_CID_END) {
-            next_cid = BLE_L2CAP_COC_CID_START;
+    if(conn) {
+    /* Get first free cid */
+        if((next_cid = ble_l2cap_coc_alloc_cid(conn->bhc_cid_mask)) == -1) {
+        	next_cid = -1;
```

Comment:
`next_cid = -1` is not needed here.
Also nitpic I would do 
```
next_cid = ble_l2cap_coc_alloc_cid(conn->bhc_cid_mask);
if (next_cid < 0) {
    goto done;
}
```

Reply:
fixed